common/timing.cc

```diff
@@ -211,9 +232,32 @@ delay_t timing_analysis(Context *ctx, bool print_fmax, bool print_path)
             log_break();
         }
     }
-    if (print_fmax)
-        log_info("estimated Fmax = %.2f MHz\n", 1e6 / (default_slack - min_slack));
-    return min_slack;
+
+    delay_t default_slack = delay_t(1.0e12 / ctx->target_freq);
+    log_info("estimated Fmax = %.2f MHz\n", 1e6 / (default_slack - min_slack));
+
+    if (print_histogram) {
+        constexpr unsigned num_bins = 20;
+        unsigned bar_width = 60;
+        auto min_slack = slack_histogram.begin()->first;
+        auto max_slack = slack_histogram.rbegin()->first;
+        auto bin_size = (max_slack - min_slack) / num_bins;
+        std::vector<unsigned> bins(num_bins + 1);
+        unsigned max_freq = 0;
+        for (const auto &i : slack_histogram) {
+            auto &bin = bins[(i.first - min_slack) / bin_size];
+            bin += i.second;
+            max_freq = std::max(max_freq, bin);
+        }
+        bar_width = std::min(bar_width, max_freq);
+
+        log_break();
+        log_info("Slack histogram:\n");
+        log_info(" legend: * represents %d endpoint(s)\n", max_freq / bar_width);
+        for (unsigned i = 0; i < bins.size(); ++i)
+            log_info("%6d < ps < %6d |%s\n", min_slack + bin_size * i, min_slack + bin_size * (i + 1),
```

Comment:
As I understand it, delay_t is not guaranteed to be an integer (indeed I don't think it is in generic) . I may have previously made the same mistaken assumption in the original code, so not really your fault.

Reply:
Good thinking batman! I'll resolve it tomorrow.

Reply:
Thanks!

Reply:
https://github.com/YosysHQ/nextpnr/pull/31/commits/8974ef332737bcf2f37cdd59b2c9767fa187fdf3 changes `DelayFrequency`'s key to be an `int` type, and uses `int(Arch::getDelayNS() * 1000)` for picosecond granularity.

That should do it, right?

Reply:
Yeah, that looks good